Update CODEOWNERS section for metadata and wildcards (#6968)
Azure/azure-sdk: docs/policies/opensource.md

{% include requirement/MUST %} edit the root-level CODEOWNERS file to ensure that it is updated to redirect all pull requests for the directory of the client library to point to the relevant engineers of this component. If the client library will exist within its own repository, then a CODEOWNERS file must be introduced and configured appropriately.

Use the following rules to ensure that we can use CODEOWNERS for both GitHub and build failure notifications:

* The CODEOWNERS file must be located in the .github directory. `/.github/CODEOWNERS`

> * CODEOWNERS has METADATA which exists in the comments of the CODEOWNERS file. The METADATA, and what it's for, are fully described [here](https://github.com/Azure/azure-sdk-tools/blob/main/tools/codeowners-utils/METADATA.md#codeowners-metadata). The METADATA is primarily used when by [GitHub Events](https://github.com/Azure/azure-sdk-tools/blob/main/tools/github-event-processor/RULES.md) to do Issue and Pull Request triage and processing.

Author: Change made locally and will be pushed with the next commit.

* Follow the `/sdk/<service name>/` (with the leading and trailing slashes) convention to define service owners
  * When using this format, service owners will be automatically subscribed to build notification failure alerts

>   * Wildcards are supported for paths but the path must end in a `/`. For example if someone owned `/sdk/foo1`, `/sdk/foo2` and `/sdk/foo3` the CODEOWNERS could have one entry with `/sdk/foo*/`. Similarly, if a user or team owned a particular file that existed in numerous subdirectories that can be wildcarded with the following `/**/ci.something.yml`. This says I own all of the `ci.something.yml` files regardless of where in they live in the repository. If this was to be limited to the `/sdk` directory it would simply be `/sdk/**/ci.something.yml`

Reviewer: ```suggestion
  * Wildcards are supported for paths but the path must end in a `/`. For example, if someone owned `/sdk/foo1`, `/sdk/foo2` and `/sdk/foo3` the CODEOWNERS could have one entry with `/sdk/foo*/`. Similarly, if a user or team owned a particular file that existed in numerous subdirectories that can be wildcarded with the following `/**/ci.something.yml`. This says I own all of the `ci.something.yml` files regardless of where in they live in the repository. If this was to be limited to the `/sdk` directory it would simply be `/sdk/**/ci.something.yml`
```

Author: Change made locally and will be pushed with the next commit.

* Place more general rules higher in the file and more specific rules lower in the file as GitHub uses the last matching expression.

>   * The exception here would be certain wildcarded rules, like `/**/ci.something.yml`. This would need to be lower in the file to get picked up correctly since GitHub uses the last matching expression. For example if `/sdk/MyService/` had a `ci.something.yml` and the `/sdk/MyService/` entry was lower than the `/**/ci.something.yml`, the owners of `/sdk/MyService/` would be identified as the owners of `ci.something.yml` because it's lower in the file.

Reviewer: ```suggestion
  * The exception to this is certain wildcarded rules, like `/**/ci.something.yml`. This would need to be lower in the file to get picked up correctly since GitHub uses the last matching expression. For example if `/sdk/MyService/` had a `ci.something.yml` and the `/sdk/MyService/` entry was lower than the `/**/ci.something.yml`, the owners of `/sdk/MyService/` would be identified as the owners of `ci.something.yml` because it's lower in the file.
```

Author: Making the changes locally, will update with the next commit.

> * Use GitHub person aliases for the owners (e.g. `@person`) and/or a GitHub team (e.g. `@Azure/team1`). GitHub teams must be within the `Azure` org. GitHub users that aren't linked to internal users, internal group aliases, email addresses and non-Azure org teams are not supported.

Reviewer: You have an example already, but it may be helpful to call out explicitly that the format for a team must start with `@Azure/`, as it has been a major point of confusion that I've seen in PRs.

Author: I'll add this as another sub bullet to sub list after line 73 and the next commit will contain the change.

>   * In order to be in CODEOWNERS file, a user or team needs to be in the Azure org and have write permission. This means they need to be part of the `azure-sdk-write` team. This also applies users and groups in METADATA tags.

Reviewer: ```suggestion
  * In order to be in CODEOWNERS file, a user or team needs to be a public member of the GitHub Azure org and have write permission. This means they need to be part of the `azure-sdk-write` team. This also applies users and groups in METADATA tags.
```

Author: Making the change locally, will push with the next commit.

* If you want PRs in those folders to be auto-labeled, add a comment line above the entry with the path with the content of `# PRLabel: ` followed by the `%Label` you want to apply. Note: Each path must its own label. Having one `# PRLabel: ` followed by multiple paths will only apply the label to the first path.

> * You can also capture the information about which people have to be notified when issues are filed for a service. To do that, you have to add the `# ServiceLabel: ` followed by the `%Label` that have to be applied to an issue in order for the people specified in the path below to be mentioned by the bot.

Reviewer: ```suggestion
* You can also capture the information about which people have to be notified when issues are filed for a service. To do that, you have to add the `# ServiceLabel: ` followed by the `%Label` that have to be applied to an issue in order for the people specified in the path below to be mentioned by the bot.  If more than one label is specified, the issue must have all of those labels applied for those people to be mentioned.
```

Author: As per our conversation, this processing doesn't exist. There's no differentiation between labels, they're just strings, and no processing done that relies on sets of labels. That might be something you want to get to, but it's not there now. Further, this isn't something that can be half done, either all CODEOWNERS need to be updated with this data prior to any rules changes or this won't ever work.

* If the code for a service is not inside the repo, you can use this special commented out path to allow issues to be tagged for a service: `# ServiceOwners: `. Note: This was previously `#/<NotInRepo>/` which was renamed to be more intuitive and tooling, for the moment, recognizes both.
* For the labels that are used with the `%` character, you can use spaces. Labels are delimited by the start of the `%` character.

> * Types of owners - With the METADATA there 3 types of owners

Reviewer: ```suggestion
* Types of owners - within the METADATA there 3 types of owners:
```

You may want to consider just using `* Types of owners:` as I'm not sure the remainder offers a huge amount of value.

Author: I went with `Types of owners:`. Will be pushed with the next commit.

>   1. Source owners - These are the user/teams on the source path lines `/sdk/azconfig/   @person3 @person4`. Th

Reviewer: ```suggestion
  1. Source owners - These are the user/teams on the source path lines `/sdk/azconfig/   @person3 @person4`. These are the individuals that GitHub will assign to review PRs automatically.
```

Author: Good catch. Will be updated with the next commit.

  2. Service owners - These are user/teams after the `# ServiceOwners: ` or `#/<NotInRepo>/` monikers. If a `# ServiceLabel: ` entry is attached to a source line, the source owners are also service owners.

>   3. AzureSdkOwners - These are owners that have triage responsibility for a given service label. Must be part of block with a `# ServiceLabel: ` entry.

Reviewer: ```suggestion
  3. AzureSdkOwners - These are owners that have triage responsibility for a given service label definition. Must be part of block with a `# ServiceLabel: ` entry.
```

The original phrasing sounds like it was limited to a single label.  Not in love with my suggestion, but I couldn't think of anything better.

Author: Will be updated with the next commit.

/sdk/**/azure-resourcemanager-*/   @Azure/team1 @person11

# A service label on a source path identifies the source owners as also being the service owners.
# ServiceLabel: %KeyVault %Service Attention
/sdk/keyvault/   @person5 @person6

# PR label for a given source path
# PRLabel: %label
/sdk/servicebus/ @person7 @person8

# Example for a service that needs both PRs to be labeled and service owners for issue triage
# ServiceLabel: %label
# PRLabel: %label
/sdk/eventhubs/ @person7 @person8

# Example for a service that needs both PRs to be labeled, service owners and AzureSdkOwners for issue triage
# ServiceLabel: %label
# PRLabel: %label
# AzureSdkOwners: @Azure/team3 @person12
/sdk/eventhubs/ @person7 @person8

# Example for service that does not have the code in the repo but wants issues to be labeled and service owners for issue triage
# ServiceLabel: %Some Label %Some Other Label
# ServiceOwners: @person7 @person8

> # Example for service that does not have the code in the repo and has service owners as well as AzureSdkOwners responsible for issue triage for the service

Reviewer: We should consider showing an example where:

- The "%label %Client" is owned by our team (code and triage) and has service owners
- The "%label %Management" is owned by others from our team (code and triage) and has different service owners.

This is a fairly common pattern that we'll see, and it would be helpful to have an example to point people at.

Author: As per our conversation, this processing doesn't exist. There's no differentiation between labels, they're just strings, and no processing done that relies on sets of labels. That might be something you want to get to, but it's not there now. Further, this isn't something that can be half done, either all CODEOWNERS need to be updated with this data prior to any rules changes or this won't ever work.

* Follow the `/sdk/<service name>/` (with the leading and trailing slashes) convention to define service owners
  * When using this format, service owners will be automatically subscribed to build notification failure alerts
  * Wildcards are supported for paths but the path must end in a `/`. For example, if someone owned `/sdk/foo1`, `/sdk/foo2` and `/sdk/foo3` the CODEOWNERS could have one entry with `/sdk/foo*/`. Similarly, if a user or team owned a particular file that existed in numerous subdirectories that can be wildcarded with the following `/**/ci.something.yml`. This says I own all of the `ci.something.yml` files regardless of where in they live in the repository. If this was to be limited to the `/sdk` directory it would simply be `/sdk/**/ci.something.yml`.
* Place more general rules higher in the file and more specific rules lower in the file as GitHub uses the last matching expression.

>   * The exception to this is certain wildcarded rules, like `/**/ci.something.yml`. This would need to be lower in the file to get picked up correctly since GitHub uses the last matching expression. For example if `/sdk/MyService/` had a `ci.something.yml` and the `/sdk/MyService/` entry was lower than the `/**/ci.something.yml`, the owners of `/sdk/MyService/` would be identified as the owners of `ci.something.yml` because it's lower in the file.

Reviewer 2: What is that an exception? It seems to follow the same rules and needs to be lower in the file.

Author: I updated the example to be more clear. The purpose was to show what would happen if `/sdk/MyService/` was lower in the file than `/**/ci.something.yml`

Reviewer 2: OK but that isn't an exception it is true about every pattern. The lowest matching pattern will be the one that wins.
